tensorflow-sys/src/runtime_linking/runtime.rs

```diff
@@ -0,0 +1,114 @@
+#[doc(hidden)]
+macro_rules! link {
+    (
+        $(
+            extern "C" {
+                pub fn $name:ident($($pname:ident: $pty:ty),* $(,)?) $(-> $ret:ty)*;
+            }
+        )+
+    ) => (
+        use lazy_static::lazy_static;
+        use libloading;
+        use std::path::PathBuf;
+        use std::sync::Arc;
+        use std::sync::RwLock;
+
+        // Wrap the loaded functions.
+        pub(crate) struct SharedLibrary {
+            library: libloading::Library,
+            path: PathBuf,
+            pub functions: Functions,
+        }
+        impl SharedLibrary {
+            fn new(library: libloading::Library, path: PathBuf) -> Self {
+                Self {
+                    library,
+                    path,
+                    functions: Functions::default(),
+                }
+            }
+        }
+
+        // `LIBRARY` holds the shared library reference.
+        lazy_static!{
+            static ref LIBRARY: RwLock<Option<Arc<SharedLibrary>>> = RwLock::new(None);
+        }
+
+        // Helper function for accessing the thread-local version of the library.
+        fn with_library<T, F>(f: F) -> Option<T>
+        where
+            F: FnOnce(&SharedLibrary) -> T,
+        {
+            match LIBRARY.read().unwrap().as_ref() {
+                Some(library) => Some(f(&library)),
+                _ => None,
+            }
+        }
+
+        // The set of functions loaded dynamically.
+        #[derive(Default)]
+        pub(crate) struct Functions {
+            $(
+                pub $name: Option<unsafe extern fn($($pname: $pty), *) $(-> $ret)*>,
+            )+
+        }
+
+        // Provide functions to load each name from the shared library into the `SharedLibrary`
+        // struct.
+        mod load {
+            $(
+                pub(crate) fn $name(library: &mut super::SharedLibrary) {
+                    let symbol = unsafe { library.library.get(stringify!($name).as_bytes()) }.ok();
+                    library.functions.$name = match symbol {
+                        Some(s) => *s,
+                        None => None,
+                    };
+                }
+            )+
+        }
+
+        /// Load all of the function definitions from a shared library.
+        pub fn load() -> Result<(), String> {
+            match crate::library::find() {
+                None => Err("Unable to find the `libtensorflow` library to load".into()),
+                Some(path) => load_from(path),
+            }
+        }
+        fn load_from(path: PathBuf) -> Result<(), String> {
+            let library = Arc::new(SharedLibrary::load(path)?);
+            *LIBRARY.write().unwrap() = Some(library);
```

Comment:
Since this is returning a `Result` anyway, this should return an `Err` on failure rather than calling `unwrap()` and panicking.

Reply:
Good point, updated to fix this.

Reply:
I misunderstood what was being asked apparently and have updated the solution.

Reply:
I meant that we shouldn't be using `unwrap()`, so instead of

```rust
let mut static_lib = LIBRARY.write().map_err(...).unwrap();
```

it should be

```rust
let mut static_lib = LIBRARY.write().map_err(...)?;
```